server/src/apub/fetcher.rs

```diff
@@ -109,50 +58,114 @@ pub fn search_by_apub_id(query: &str, conn: &PgConnection) -> Result<SearchRespo
   };
   match fetch_remote_object::<SearchAcceptedObjects>(&query_url)? {
     SearchAcceptedObjects::Person(p) => {
-      let u = upsert_user(&UserForm::from_person(&p)?, conn)?;
-      response.users = vec![UserView::read(conn, u.id)?];
+      let user_uri = p.base.base.object_props.get_id().unwrap().to_string();
+      let user = get_or_fetch_and_upsert_remote_user(&user_uri, &conn)?;
+      response.users = vec![UserView::read(conn, user.id)?];
     }
     SearchAcceptedObjects::Group(g) => {
-      let c = upsert_community(&CommunityForm::from_group(&g, conn)?, conn)?;
-      fetch_community_outbox(&c, conn)?;
-      response.communities = vec![CommunityView::read(conn, c.id, None)?];
-    }
-    SearchAcceptedObjects::Page(p) => {
-      let p = upsert_post(&PostForm::from_page(&p, conn)?, conn)?;
-      response.posts = vec![PostView::read(conn, p.id, None)?];
+      let community_uri = g.base.base.object_props.get_id().unwrap().to_string();
+      let community = get_or_fetch_and_upsert_remote_community(&community_uri, &conn)?;
+      // TODO Maybe at some point in the future, fetch all the history of a community
+      // fetch_community_outbox(&c, conn)?;
+      response.communities = vec![CommunityView::read(conn, community.id, None)?];
     }
   }
   Ok(response)
 }
 
-/// Fetch all posts in the outbox of the given user, and insert them into the database.
-fn fetch_community_outbox(community: &Community, conn: &PgConnection) -> Result<Vec<Post>, Error> {
-  let outbox_url = Url::parse(&community.get_outbox_url())?;
-  let outbox = fetch_remote_object::<OrderedCollection>(&outbox_url)?;
-  let items = outbox.collection_props.get_many_items_base_boxes();
-
-  Ok(
-    items
-      .unwrap()
-      .map(|obox: &BaseBox| -> Result<PostForm, Error> {
-        let page = obox.clone().to_concrete::<Page>()?;
-        PostForm::from_page(&page, conn)
-      })
-      .map(|pf| upsert_post(&pf?, conn))
-      .collect::<Result<Vec<Post>, Error>>()?,
-  )
+/// Check if a remote user exists, create if not found, if its too old update it.Fetch a user, insert/update it in the database and return the user.
+pub fn get_or_fetch_and_upsert_remote_user(
```

Comment:
These are some neat methods I added, based on grishkas comments about fetching users, and why I recently added a `last_checked_at` column to the user and community tables.

It goes:

- We receive an activity.
- If we don't have that user or community, fetch it, and insert.
- If we do, see if its older than a certain time.
  - If its old, re-fetch, update it, and update a last_checked to now.
  - If its new, just read it from the DB.

Reply:
Why dont you handle this with an optional parameter for `fetch_remote_object()` like I suggested? Seems simpler to me, you just pass last_updated (or None) into that function, and it checks if the fetch is needed. In that case it would have to return the object from the database so you would have to pass that in as well. Should simplify the code here a lot.

Reply:
Whats the optional param? It would also need to know the type of object its parsed into.

Reply:
The updated_time. So basically it would be `fetch_remote_object(url, object_from_database, Some(updated_time))`.

Reply:
I couldn't find a good way to do this, because it requires when (after checking the time), updating that DB row, which could be either a user or a community row.

Reply:
Hmm yeah thats true.